### ISSUE TITLE
build_engine(fp16=True) silently builds an FP32 engine on TensorRT 11

### ISSUE THREAD

**Issue by lthphuw** [roboflow/rf-detr#1453]
### Search before asking

- [x] I have searched the RF-DETR issues and found no similar bug report.

### Bug

> **Related to #1214, filed deliberately rather than as an oversight.** #1214 reported this same
> breakage and was closed on 2026-07-29 with *"if you still observe these issues, pls feel free to
> reopen it"*. I am not its author, so I cannot reopen it — this is a fresh report taking up that
> invitation, with measurements on an L4 and proof of the engine's actual precision. Glad to have
> this closed as a duplicate if you would rather reopen #1214 instead. Details in *Additional*.

**Executed Colab notebook with all outputs (NVIDIA L4):**
https://colab.research.google.com/drive/1wRDsTbAV800_gFZHzyjP5bUm7tAy7YWO?usp=sharing

### What happens

`pip install rfdetr[tensorrt]` resolves to TensorRT 11 today — the extra declares `tensorrt>=8.6.1`
with no upper bound. On TensorRT 11, `build_engine(..., fp16=True)` logs one warning, **exits 0, and
writes an FP32 engine**. Nothing raises. The engine loads, runs, and can be benchmarked — and the
number is an FP32 number.

Measured on an **NVIDIA L4**, `rfdetr` **1.10.1**, `rfdetr-small` @ 512², batch 1, TensorRT 11.2.1.2.
Engine forward only, 50 warmup + 300 CUDA-event-timed runs:

| Engine | Bytes | % layer I/O tensors in Half | mean ms |
|---|---|---|---|
| `build_engine(fp16=True)` | 119,688,644 | **0.00** | **8.407** |
| same graph cast to FP16, built strongly typed | 65,059,180 | 86.87 | **2.170** |

**3.87× slower** than the FP16 the caller asked for. Anyone benchmarking RF-DETR against other
detectors publishes an FP32 latency labelled FP16.

Build log — the downgrade is visible, but non-fatal:

```
[WARNING] rf-detr - TensorRT 11.2.1.2 does not expose the FP16 builder flag; building an FP32 engine instead. Pass fp16=False to silence this warning.
[INFO]    rf-detr - Building TensorRT engine (fp16=False) from /content/export/rfdetr-small.onnx
    Flags | [STRONGLY_TYPED]
[I] Finished engine building in 19.516 seconds
[I] Saving engine to /content/export/rfdetr-small_fp32.trt     <- exit 0
```

Credit where due: the `_fp32` suffix means the artifact on disk is honestly named. The problem is
that `fp16=True` was requested and silently not delivered.

### Root cause

TensorRT 11 removed weak typing. `BuilderFlag.FP16`, `BuilderFlag.INT8`,
`Builder.platform_has_fast_fp16`, `IInt8Calibrator` and `NetworkDefinitionCreationFlag.EXPLICIT_BATCH`
are all gone. Every network is strongly typed, and **precision is taken from the dtypes in the ONNX
graph**, not from builder flags. Verified on the L4:

```python
tensorrt: 11.2.1.2
BuilderFlag.FP16: False
platform_has_fast_fp16: False
network flags: ['PREFER_AOT_PYTHON_PLUGINS', 'PREFER_JIT_PYTHON_PLUGINS', 'STRONGLY_TYPED']
```

`src/rfdetr/export/_tensorrt.py:132` (line 132 in the released 1.10.1 wheel) probes for the flag and
downgrades when it is absent:

```python
if not hasattr(trt.BuilderFlag, "FP16"):
    logger.warning(
        "TensorRT %s does not expose the FP16 builder flag; building an FP32 engine instead. "
        "Pass fp16=False to silence this warning.", ...)
    fp16 = False
    engine_path = _engine_path(fp16_used=fp16)
```

Per its own comment, that probe exists for **lean/partial TensorRT wheels** that do not expose the
flag. On TensorRT 11 it fires for an entirely different reason — the flag was *deleted from the API* —
and the safe fallback becomes a permanent, silent loss of the requested precision.

The other half of the cause sits upstream of the builder: **the exported ONNX is 100 % float32.**
`export_onnx` performs no cast, and `RFDETR.export`'s `fp16` argument is documented as TensorRT-only.
So under strong typing there is nothing for TensorRT to build in FP16 even in principle. Verified on
the L4: `{'FLOAT': 354}` initializers, zero FLOAT16, and all three graph I/O tensors FLOAT.

Worth noting: simply deleting the guard is *not* the fix. polygraphy then fails loudly instead —
`PolygraphyException: fp16 in CreateConfig is not available on TensorRT version 11.2.1.2`. The guard
is precisely what converts a hard failure into a silent FP32 engine.

### How the FP32-ness is proven

A successful build is not evidence of precision, so this is not argued from the filename or the flag.
TensorRT 11's engine inspector omits the per-layer `Precision` field, so the probe histograms
per-tensor datatypes across every layer's inputs and outputs (this needs
`profiling_verbosity=DETAILED`, which polygraphy sets by default).

The probe is **calibrated in the same notebook, on the same L4**, against two engines of identical
topology that differ only in graph dtype:

| Probe input | reported % Half |
|---|---|
| known-FP32 graph, strongly typed | **0.00** |
| known-FP16 graph, strongly typed | **100.00** |

Applied to what `build_engine(fp16=True)` produced: **0.00 % Half**,
`{'Float': 791, 'Bool': 2, 'Int32': 4}`.

Two independent cross-checks agree — the serialized engine is ~1.84× the size of a real FP16 build,
and the forward is 3.87× slower.

(Exact serialized byte counts drift by <0.05 % between builds because TensorRT picks tactics
non-deterministically; the two figures that do not move are `0.00 %` and `86.87 %`.)

### Environment

Reproduced on the **latest PyPI release**, not just a dev branch:

- **RF-DETR: 1.10.1** (installed by a plain `pip install rfdetr` in Colab on 2026-09-10). The
  `hasattr(trt.BuilderFlag, "FP16")` guard is present at `rfdetr/export/_tensorrt.py:132` in the
  published `rfdetr-1.10.1-py3-none-any.whl`.
- Also present on `develop` @ `75c2503` (1.11.0.dev0), and reported earlier against 1.9.4 in #1214.
- OS / Python: Google Colab, Python 3.13.15
- PyTorch: 2.11.0+cu128 (CUDA 12.8, cuDNN 91900)
- **TensorRT: 11.2.1.2** (`tensorrt_cu13`) — contrast case **10.16.1.11**, where the FP16 flag still
  exists and the engine really is FP16
- GPU: **NVIDIA L4**, compute capability 8.9, 23034 MiB

Second machine, for the same bug on a different model/resolution/CUDA line — RTX 3060 (compute 8.6),
TensorRT 11.2.1.2, `rfdetr-medium` @ 704²: `build_engine(fp16=True)` gives **0.00 % Half** and
26.26 ms; the cast graph gives 88.76 % Half and 5.75 ms.

Exact version line from the notebook:

```
rfdetr 1.10.1 | torch 2.11.0+cu128 | cuda 12.8 | cudnn 91900 | trt 11.2.1.2 | py 3.13.15 | NVIDIA L4
```

Note the TensorRT wheel is the CUDA 13 build (`tensorrt_cu13`) while PyTorch is built against
CUDA 12.8 — TensorRT ships its own CUDA libraries, so this mix is expected and not the cause.

### Minimal Reproducible Example

**Executed notebook with outputs (NVIDIA L4):**
https://colab.research.google.com/drive/1wRDsTbAV800_gFZHzyjP5bUm7tAy7YWO?usp=sharing

Self-contained and **needs no dataset** — the bug is in the ONNX→engine precision plumbing and is
data-independent, so the notebook uses stock COCO-pretrained weights and random input tensors. Every
step ends in an `assert`, including a calibration step that verifies the precision probe reports
0.00 % on a known-FP32 engine and 100.00 % on a known-FP16 one before it is trusted on the real one.

Minimal version:

```bash
pip install rfdetr onnx "tensorrt==11.2.1.2" polygraphy
```

```python
import tensorrt as trt
print(trt.__version__, hasattr(trt.BuilderFlag, "FP16"))
# 11.2.1.2 False

from rfdetr import RFDETRSmall
from rfdetr.export._tensorrt import build_engine

onnx_path = str(RFDETRSmall().export(output_dir="/content/export", format="onnx", verbose=False))
engine = build_engine(onnx_path, fp16=True)     # one warning, exit code 0
print(engine)
# /content/export/rfdetr-small_fp32.trt   <- FP32, despite fp16=True
```

To confirm the engine really is FP32 rather than trusting the filename, see section 4 of the
notebook (per-layer datatype histogram via `IEngineInspector`, calibrated in section 4b).

### A real FP16 engine *is* achievable on TensorRT 11

Cast the ONNX graph to FP16 first, then build strongly typed. Two casters were measured; **both
work, and neither hits an unsupported operator** — `GridSample` (deformable attention) converts fine.

| | `onnxconverter-common` | ModelOpt AutoCast (NVIDIA's documented path) |
|---|---|---|
| package size | **428 KB** | ~350 MB of genuinely new deps (cupy, scipy, sympy, pulp) |
| cast time | **2.0 s** | 43 s |
| engine (3060, medium @ 704²) | 72,046,092 B, **88.76 % Half**, 5.75 ms | 72,499,668 B, 87.95 % Half, 5.79 ms |
| IoU vs FP32 (min / mean) | **0.9761** / 0.9957 | 0.9716 / 0.9956 |

`onnxconverter-common` looks like the better trade: its only dependencies (`onnx`, `numpy`,
`protobuf`) are already rfdetr dependencies, so the marginal install cost is effectively zero.

It does need two small fix-ups, both about ten lines, and both found only because the build failed:

1. It does not rewrite the `to` attribute of `Cast` nodes that already exist in the source graph.
   RF-DETR exports 33–35 `Cast(to=FLOAT)` nodes, so the tensor stays float32 while its `value_info`
   claims float16, and TensorRT rejects the graph at the first Conv
   (`INVALID_NODE ... patch_embeddings/projection/Conv`).
2. Its `keep_io_types=True` leaves the FP32 graph input wired straight into an FP16 Conv without
   inserting a Cast, producing the same rejection.

Notably, `onnx.checker.check_model` **passes** on both malformed graphs — only TensorRT's
strongly-typed parser catches them.

**Accuracy cost of FP16** (RTX 3060, 32 real frames, versus ONNX Runtime CPU FP32 on the uncast
graph, boxes matched **by IoU**): 309/309 boxes recovered, 0 unmatched at IoU 0.5, IoU min 0.976 /
mean 0.996; 3/309 flip across a 0.4 score threshold.

> Measurement note for anyone re-running this: comparing `dets[i]` against `ref_dets[i]` *slot-wise*
> reports a 142.9 px deviation. That is FP16 `TopK` reordering the query slots, not box drift. Match
> by IoU before drawing any accuracy conclusion.

### Relation to #1214

#1214 described this correctly and was closed on 2026-07-29 with *"I believe this was resolved in
#853 and #1242 … However, if you still observe these issues, pls feel free to reopen it"*. I am not
the author of #1214 so I cannot reopen it — this is a fresh report taking up that invitation, now
with measurements on an L4. Happy to close this as a duplicate if you would rather reopen #1214.

For the record, on the two PRs cited there: #853 added the polygraphy export path, and #1242 is
`fix(deps): require onnxsim>=0.7.0 …`, unrelated to TensorRT. The `hasattr` FP16 guard actually
arrived in **#1231** (merged 2026-07-24). #1231 fixed a real problem — the hard crash on TensorRT 11 —
so this is not a bad fix; the precision loss it introduced simply went unnoticed.

Two things help explain why it stayed unnoticed:

- **The cookbook can't surface it.** `docs/cookbooks/export-tensorrt.ipynb` installs
  `inference-models[trt10]`, whose extra pins `tensorrt-cu12>=10.0.0,<11.0.0` (PyPI metadata,
  `inference-models` 0.37.0). The cookbook is therefore forced onto TensorRT 10, where FP16 genuinely
  works. It also never asserts precision — it prints the engine size and runs inference, and both
  succeed on an FP32 engine. Its text currently states "The engine is FP16 by default (lowest
  latency); it automatically falls back to FP32 (with a warning) if your TensorRT build doesn't
  expose the FP16 builder flag", which on TensorRT 11 is unconditional.
- **A test asserts the current behaviour is correct.**
  `tests/export/test_tensorrt_export.py::TestBuildEngineFp16Fallback::test_downgrades_to_fp32_when_fp16_flag_unavailable`
  asserts the silent FP32 downgrade, using a fake `__version__ = "11.1.0.106"`. And no test asserts
  that an FP16 request yields an FP16 engine — the GPU e2e test builds with `fp16=False`. From inside
  CI, the behaviour looks correct.

### Separately broken on TensorRT 11: `export/benchmark.py`

`TRTInference` uses six symbols removed in TensorRT 10/11. Four are on the *inference* path, and
since `run_async` is the default (`sync_mode=False`), the class is unusable on TensorRT 11 even for a
pre-built engine:

Line numbers below are on `develop` @ `75c2503`:

| Line | Symbol | Removed in |
|---|---|---|
| 420 | `NetworkDefinitionCreationFlag.EXPLICIT_BATCH` | TensorRT 10 |
| 428 | `BuilderFlag.FP16` | TensorRT 11 |
| 344, 351, 363 | `enumerate(engine)` (`ICudaEngine.__iter__`) | TensorRT 11 |
| 387 | `context.execute_async_v2` | TensorRT 11 |

This one was already broken on TensorRT 10 and is gated behind `[tensorrt-bench]` + `pycuda`, so it
is probably a separate issue — happy to file it if useful.

### Suggested direction

Listing these separately because they are separable, and because two of them are your call rather
than mine:

1. Distinguish "flag absent because lean wheel" from "flag absent because TensorRT ≥ 11 strong
   typing", and never satisfy an `fp16=True` request with an FP32 engine silently.
2. On TensorRT ≥ 11, cast the ONNX and build strongly typed. **This adds a dependency**
   (`onnxconverter-common`) — your call whether that belongs in `[tensorrt]` or a new extra.
3. Raise an actionable error when FP16 is requested and cannot be delivered, naming `tensorrt<11` as
   the fallback. **This is a behaviour change** for anyone currently on TensorRT 11.
4. Replace the test that asserts the downgrade, and add a GPU e2e test that proves FP16 via the
   calibrated probe rather than just checking the file exists.

I have the fix, tests and notebook ready and am glad to open the PR — but points 2 and 3 are
maintainer decisions, so I would rather get a nod on the approach first.

### Are you willing to submit a PR?

- [x] Yes, I'd like to help by submitting a PR!

**Comment by modelpath-dev** [comment]
I will take this issue. Please assign it to me.

The problem is that `build_engine(fp16=True)` does not raise an error when TensorRT 11 silently builds an FP32 engine. I will first check the code where `build_engine` is defined. The goal is to ensure it raises an exception or logs a clear error if `fp16=True` is requested but not supported. I will also verify the TensorRT version check to ensure it correctly identifies unsupported features. The fix will likely involve adding a version check and raising an error if FP16 is not available.

**Comment by lthphuw** [comment]
@Borda — I opened this issue and already have a working branch for it — tests passing on CPU, no public API change. Posting the approach here before a PR in case you'd rather steer it differently. @modelpath-dev, note the direction below goes further than erroring out: it actually builds the FP16 engine on TRT 11.

**Root cause.** TensorRT 11 is strongly typed: `BuilderFlag.FP16` is gone and precision comes from the ONNX graph. `build_engine(fp16=True)` sets no flag and returns an FP32 engine named `_fp16.trt` — measured 3.87x slower than the FP16 that was requested.

**Approaches I evaluated**

| Approach | Verdict |
| --- | --- |
| Raise / warn + downgrade to FP32 | Rejected — makes the bug visible but leaves TRT 11 with no FP16 at all |
| Hand-roll the cast on plain `onnx`, per the #1054 GridSample precedent | Rejected — my 28-line prototype passes `onnx.checker` but fails strict type inference (`Inferred elem type differs: (10) vs (1)`), exactly what the TRT 11 parser rejects. A correct version is ~150-250 LOC (block list, boundary casts, subgraph recursion). #1054 also uses `onnx_graphsurgeon`, which `[tensorrt]` does not ship. |
| **Cast the graph with `onnxconverter-common`** &larr; recommended | 87 KiB pure-Python wheel, **zero** new transitive packages (`numpy`/`packaging`/`protobuf` are already in the extra). polygraphy's own `convert_to_fp16` is a thin wrapper over this same library. |

`onnx` (~19 MB) is the only real new weight, and it is unavoidable for any graph-level approach — `polygraphy` declares no requirements, so `[tensorrt]` never pulled it in.

**Scope of the edit**

- `export/_tensorrt.py` — cast the graph on TRT >= 11, keeping FP32 graph I/O so callers are unaffected. TRT < 11 keeps the existing flag path and the existing FP32-downgrade warning unchanged. The cast file is a build intermediate and is deleted after the build.
- `export/benchmark.py` — `TRTInference.build_engine` calls `set_flag(trt.BuilderFlag.FP16)` unconditionally and hard-fails with `AttributeError` on TRT 11. Same root cause; happy to split this into its own PR.
- `pyproject.toml` — the two deps above. Also one line on the `ml-dtypes` override: a uv override replaces a package's whole requirement set, so the lone 3.12-scoped entry leaves `ml-dtypes` uninstalled elsewhere and `import onnx` then fails on 3.10/3.11/3.13.
- `tests/export/test_tensorrt_export.py` — CPU-only; the cast is pure ONNX surgery, so it runs without TensorRT.

No public API change.

**Comment by modelpath-dev** [comment]
Sounds like you've got a solid approach there. Let me know if there's anything specific you need from my side to help move this forward.
